controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,410 @@
+	for _, ip := range diff(pool.Status.IPs, pool.Spec.PreAllocatedFloatingIPs) {
+		if err := networkingService.DeleteFloatingIP(pool, ip); err != nil {
+			return fmt.Errorf("delete floating IP: %w", err)
+		}
+	}
+
+	if controllerutil.RemoveFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+		log.Info("Removing finalizer from OpenStackFloatingIPPool")
+		return r.Client.Update(ctx, pool)
+	}
+	return nil
+}
+
+func union(a []string, b []string) []string {
+	m := make(map[string]struct{})
+	for _, item := range a {
+		m[item] = struct{}{}
+	}
+	for _, item := range b {
+		m[item] = struct{}{}
+	}
+	result := make([]string, 0, len(m))
+	for item := range m {
+		result = append(result, item)
+	}
+	return result
+}
+
+func diff(a []string, b []string) []string {
+	m := make(map[string]struct{})
+	for _, item := range a {
+		m[item] = struct{}{}
+	}
+	for _, item := range b {
+		delete(m, item)
+	}
+	result := make([]string, 0, len(m))
+	for item := range m {
+		result = append(result, item)
+	}
+	return result
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) setIPStatuses(ctx context.Context, pool *infrav1.OpenStackFloatingIPPool) error {
+	ipAddresses := &ipamv1.IPAddressList{}
+	if err := r.Client.List(ctx, ipAddresses, client.InNamespace(pool.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return err
+	}
+	pool.Status.ClaimedIPs = []string{}
+	for _, ip := range ipAddresses.Items {
+		pool.Status.ClaimedIPs = append(pool.Status.ClaimedIPs, ip.Spec.Address)
+	}
+	pool.Status.IPs = append([]string{}, pool.Spec.PreAllocatedFloatingIPs...)
+
+	pool.Status.IPs = union(pool.Status.IPs, pool.Status.ClaimedIPs)
+	pool.Status.AvailableIPs = diff(diff(pool.Status.IPs, pool.Status.ClaimedIPs), pool.Status.FailedIPs)
+	return nil
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) getIP(scope scope.Scope, pool *infrav1.OpenStackFloatingIPPool) (string, error) {
+	var ip string
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		scope.Logger().Error(err, "Failed to create networking service")
+		return "", err
+	}
+
+	if len(pool.Status.AvailableIPs) > 0 {
+		ip = pool.Status.AvailableIPs[0]
+		pool.Status.AvailableIPs = pool.Status.AvailableIPs[1:]
+		pool.Status.ClaimedIPs = append(pool.Status.ClaimedIPs, ip)
+	}
+
+	if ip != "" {
+		fp, err := networkingService.GetFloatingIP(ip)
+		if err != nil {
+			return "", fmt.Errorf("get floating IP: %w", err)
+		}
+		// If the IP exist return it, else we continue and try to allocate it if we fail to allocate it we will mark it as failed
+		if fp != nil {
+			return fp.FloatingIP, nil
+		}
+	}
+
+	// ip could be empty meaning we want to get a new one or the IP
+	// if ip is not empty it got and ip from availableIPs that does not exist in openstack
+	// we try to allocate it, if we fail we mark it as failed and skip it next time
+	fp, err := networkingService.CreateFloatingIPForPool(pool, ip)
```

Comment:
This will only work with OpenStack admin credentials. I don't see `CreateFloatingIPForPool` called with `""` as the second parameter. This feels wrong, I think we want this to be able to reserve a new IP automatically, even if it's Neutron choosing the address.